x/ibchooks/marker_hooks.go

```diff
@@ -0,0 +1,235 @@
+		Name:        chainID + "/" + data.Denom,
+		Display:     chainID + "/" + data.Denom,
+		Description: data.Denom + " from chain " + chainID,
+	}
+	return h.MarkerKeeper.SetDenomMetaData(ctx, markerMetadata, authtypes.NewModuleAddress(types.ModuleName))
+}
+
+// GetChainID returns the source chain id from packet for `07-tendermint` client connection or returns `unknown`
+func (h MarkerHooks) GetChainID(ctx sdktypes.Context, packet exported.PacketI, ibcKeeper *ibckeeper.Keeper) string {
+	chainID := "unknown"
+	channel, found := ibcKeeper.ChannelKeeper.GetChannel(ctx, packet.GetSourcePort(), packet.GetSourceChannel())
+	if !found {
+		return chainID
+	}
+	connectionEnd, found := ibcKeeper.ConnectionKeeper.GetConnection(ctx, channel.ConnectionHops[0])
+	if !found {
+		return chainID
+	}
+	clientState, found := ibcKeeper.ClientKeeper.GetClientState(ctx, connectionEnd.GetClientID())
+	if !found {
+		return chainID
+	}
+	if clientState.ClientType() == "07-tendermint" {
+		tmClientState, ok := clientState.(*tendermintclient.ClientState)
+		if ok {
+			chainID = tmClientState.ChainId
+		}
+	}
+	return chainID
+}
+
+// ResetMarkerAccessGrants removes all current access grants from marker and adds new transfer grants for transferAuths
+func ResetMarkerAccessGrants(transferAuths []sdktypes.AccAddress, marker markertypes.MarkerAccountI) error {
+	for _, currentAuth := range marker.GetAccessList() {
+		if err := marker.RevokeAccess(currentAuth.GetAddress()); err != nil {
+			return err
+		}
+	}
+	for _, transfAuth := range transferAuths {
+		if err := marker.GrantAccess(markertypes.NewAccessGrant(transfAuth, markertypes.AccessList{markertypes.Access_Transfer})); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ProcessMarkerMemo extracts the list of transfer auth address from marker part of packet memo
+func ProcessMarkerMemo(memo string) ([]sdktypes.AccAddress, markertypes.MarkerType, error) {
+	found, jsonObject := jsonStringHasKey(memo, "marker")
+	if !found {
+		return []sdktypes.AccAddress{}, markertypes.MarkerType_Coin, nil
+	}
+	jsonBytes, err := json.Marshal(jsonObject["marker"])
+	if err != nil {
+		return nil, markertypes.MarkerType_Unknown, err
+	}
+
+	var markerMemo types.MarkerPayload
+	err = json.Unmarshal(jsonBytes, &markerMemo)
+	if err != nil {
+		return nil, markertypes.MarkerType_Unknown, err
+	}
+	if markerMemo.TransferAuths == nil {
+		return []sdktypes.AccAddress{}, markertypes.MarkerType_Coin, nil
+	}
+
+	transferAuths := make([]sdktypes.AccAddress, len(markerMemo.TransferAuths))
+	for i, addr := range markerMemo.TransferAuths {
+		accAddr, err := sdktypes.AccAddressFromBech32(addr)
+		if err != nil {
+			return nil, markertypes.MarkerType_Unknown, err
+		}
+		transferAuths[i] = accAddr
+	}
+	return transferAuths, markertypes.MarkerType_RestrictedCoin, nil
+}
+
+// ProcessMarkerMemoFn processes a ics20 packets memo part to have `marker` setup information for receiving chain
+func (h MarkerHooks) ProcessMarkerMemoFn(
```

Comment:
This name is really similar to the previous one. I would recommend changing one of them.